### DISCUSSION TITLE
What's the difference between User Engagement Data and "My code snippets for product improvements"

### DISCUSSION THREAD

**Discussion by ghost** [community/community#50803] [Copilot Conversations] [answered]
In a personal account settings, there is a checkbox "Allow GitHub to use my code snippets for product improvements".
But also FAQ says "[User Engagement Data (which includes pseudonymous identifiers and general usage data) ... will continue to be collected, processed, and shared with Microsoft and OpenAI as you use GitHub Copilot.](https://github.com/features/copilot#pricing:~:text=How%20can%20users%20of%20Copilot%20for%20Individuals%20control%20use%20of%20their%20Code%20Snippets%20Data)" 

So, if i uncheck "Allow GitHub to use my snippets" does it means that no piece of my code will be stored somewhere in copilot servers? Or it is about something else?

What is "pseudonymous identifiers" is it different from "my code snippets" ?

**Comment by Akash1134** [comment, marked as answer]
Hi @ArtemRevaEpam ,

Thank you for asking this question. 

User engagement data is used by GitHub to provide the service and to enable improvements which may include:

- Evaluating GitHub Copilot, for example, by measuring the positive impact it has on the user
- Fine tuning ranking and sorting algorithms and prompt crafting
- Detecting potential abuse of GitHub Copilot or violation of [Acceptable Use Policies](https://github.com/github/docs/blob/main/site-policy/acceptable-use-policies/github-acceptable-use-policies).
- Conducting experiments and research related to developers and their use of developer tools and services.

The code you're editing would be classed as Code Snippets Data and wouldn't be collected as part of general usage data. 

> What is "pseudonymous identifiers" is it different from "my code snippets" ?

This is [what we mean](https://csrc.nist.gov/glossary/term/pseudonymous_identifier) by pseudonymous identifiers. 

Hope that helps!

**Reply by ghost** [reply]
@Akash1134 Thank you for detailed answer.
Let me please clarify. 
So under "Copilot for Individuals" plan, i can uncheck "Allow GitHub to use my code snippets for product improvements" and no piece of my code will be stored anywhere in Copilot servers? 
And so I can work on highly secured projects using copilot without fear that my code will land somewhere else?

**Reply by Akash1134** [reply]
Yes @ArtemRevaEpam Once the option to not allow GitHub to use my code snippets is unchecked, your data will not be collected as part of general usage data.

By default, when processing Code Snippets Data, GitHub takes the protection measures described in [How is the transmitted Code Snippets data protected?](https://github.com/features/copilot/#how-is-the-transmitted-code-snippets-data-protected) and follows responsible practices in accordance with our [Privacy Statement](https://docs.github.com/en/site-policy/privacy-policies/github-privacy-statement) so that the use of your telemetry data to improve these models does not result in this data being shared with other GitHub Copilot users.

You can be rest assured of data privacy as we follow responsible practices in accordance with our [Privacy Statement](https://docs.github.com/en/site-policy/privacy-policies/github-privacy-statement) to ensure that your code snippets will not be used as suggested code for other users of GitHub Copilot.

**Reply by ghost** [reply]
Perfect! Thank you @Akash1134

**Reply by Dentrax** [reply]
Hey @Akash1134, could you please clarify what you mean by saying _"your data will not be collected as part of general usage data"_? @ArtemRevaEpam says that _"will my code be stored anywhere in Copilot servers?"_. By meaning of `collected` in this context, is GitHub claims that "the code" is **NOT** stored anywhere in Copilot servers?

CoPilot for Business, at section [Code Snippets Data](https://docs.github.com/en/site-policy/privacy-policies/github-copilot-for-business-privacy-statement#code-snippets-data) says that:

> _Code snippets data is only transmitted in real-time to return Suggestions, and is discarded once a Suggestion is returned. Copilot for Business does not retain any Code Snippets Data._

... whereas the [Privacy – Copilot for Individuals](https://github.com/features/copilot/#how-is-the-transmitted-code-snippets-data-protected) does NOT mention that (retain policy in case we uncheck the `Allow GitHub to use my code snippets for product improvements` option).

Can you please elaborate a bit?

**Reply by Akash1134** [reply]
Hi @Dentrax , 

Access is strictly controlled for Copilot for Individuals to ensure there are no privacy concerns, we provide the freedom to our users of Copilot for Individuals to control the use of their Code Snippets.
- User Engagement Data (which includes pseudonymous identifiers and general usage data), is required for the use of GitHub Copilot and will continue to be collected, processed, and shared with Microsoft and OpenAI as you use GitHub Copilot.
- Users of Copilot for Individuals can choose whether Code Snippets Data is retained by GitHub and further processed and shared with Microsoft and OpenAI by [adjusting user settings](https://github.com/settings/copilot).
- Users of Copilot for Individuals can request deletion of Code Snippet Data associated with their GitHub identity. GitHub Copilot is powered by Codex, which is a learning model trained by OpenAI. To request removal of your code from Codex's training data, please follow the procedure outlined in the "Copyright Complaints" section in OpenAI's [terms of use](https://openai.com/terms/). For more information, please contact us by [filling out a support ticket](https://support.github.com/request).



The data can only be accessed by (1) named GitHub personnel working on the GitHub Copilot team or on the GitHub platform health team, (2) Microsoft personnel working on or with the GitHub Copilot team, and (3) OpenAI personnel who work on GitHub Copilot

**Reply by Dentrax** [reply]
Thanks for explanations, @Akash1134!

My understanding from your comments is that _"For Individuals, code snippets WILL be STORED in Copilot servers without applying ANY retain policy."_.

**Reply by Akash1134** [reply]
@Dentrax I'd recommend that you read the [Privacy FAQs](https://github.com/features/copilot/#:~:text=Privacy%20%E2%80%93%20Copilot%20for%20Individuals) for more details and specifics on your question.

**Reply by ghost** [reply]
> Thanks for explanations, @Akash1134!
> 
> My understanding from your comments is that _"For Individuals, code snippets WILL be STORED in Copilot servers without applying ANY retain policy."_.

@Dentrax i guess you quote "by default" policy for CodeSnippets, which is not applied in case user uncheck "Allow GitHub to use my snippets". My guess - this checkbox do exactly what it says, and support stuff approved in this thread.
